lib/integrations/awsoidc/externalcloudaudit_iam_config.go

```diff
@@ -0,0 +1,135 @@
+// limitations under the License.
+
+package awsoidc
+
+import (
+	"context"
+	"net/url"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+	"github.com/gravitational/trace"
+
+	awslib "github.com/gravitational/teleport/lib/cloud/aws"
+	"github.com/gravitational/teleport/lib/config"
+)
+
+// ConfigureExternalCloudAuditClient is an interface for the AWS client methods
+// used by ConfigureExternalCloudAudit.
+type ConfigureExternalCloudAuditClient interface {
+	PutRolePolicy(context.Context, *iam.PutRolePolicyInput, ...func(*iam.Options)) (*iam.PutRolePolicyOutput, error)
+	GetCallerIdentity(context.Context, *sts.GetCallerIdentityInput, ...func(*sts.Options)) (*sts.GetCallerIdentityOutput, error)
+}
+
+// DefaultConfigureExternalCloudAuditClient wraps an iam and sts client to
+// implement ConfigureExternalCloudAuditClient.
+type DefaultConfigureExternalCloudAuditClient struct {
+	Iam *iam.Client
+	Sts *sts.Client
+}
+
+// PutRolePolicy adds or updates an inline policy document that is embedded in
+// the specified IAM role.
+func (d *DefaultConfigureExternalCloudAuditClient) PutRolePolicy(ctx context.Context, input *iam.PutRolePolicyInput, opts ...func(*iam.Options)) (*iam.PutRolePolicyOutput, error) {
+	return d.Iam.PutRolePolicy(ctx, input, opts...)
+}
+
+// GetCallerIdentity returns details about the IAM user or role whose
+// credentials are used to call the operation.
+func (d *DefaultConfigureExternalCloudAuditClient) GetCallerIdentity(ctx context.Context, input *sts.GetCallerIdentityInput, opts ...func(*sts.Options)) (*sts.GetCallerIdentityOutput, error) {
+	return d.Sts.GetCallerIdentity(ctx, input, opts...)
+}
+
+// ConfigureExternalCloudAudit attaches an IAM policy with necessary permissions
+// for the ExternalCloudAudit feature to an existing IAM role associated with an
+// AWS OIDC integration.
+func ConfigureExternalCloudAudit(
+	ctx context.Context,
+	clt ConfigureExternalCloudAuditClient,
+	params *config.IntegrationConfExternalCloudAuditIAM,
+) error {
+	policyCfg := &awslib.ExternalCloudAuditPolicyConfig{
+		Region:              params.Region,
+		AthenaWorkgroupName: params.AthenaWorkgroup,
+		GlueDatabaseName:    params.GlueDatabase,
+		GlueTableName:       params.GlueTable,
+	}
+
+	// Given a uri for an s3 bucket with an optional prefix (folder), we need an
+	// s3 wildcard ARN to match all objects in that bucket.
+	for _, i := range []struct {
+		name   string
+		uri    string
+		arnPtr *string
+	}{
+		{"audit events", params.AuditEventsURI, &policyCfg.AuditEventsARN},
+		{"session recordings", params.SessionRecordingsURI, &policyCfg.SessionRecordingsARN},
+		{"athena results", params.AthenaResultsURI, &policyCfg.AthenaResultsARN},
+	} {
+		var err error
+		*i.arnPtr, err = s3URIToObjectWildcardARN(i.uri)
+		if err != nil {
+			return trace.Wrap(err, "parsing %s URI", i.name)
+		}
+	}
```

Comment:
:nail_care: 
This code seems a bit too clever, I would rather have the usual assignments inline